examples/hif2a/fit_to_multiple_rbfes.py

```diff
@@ -0,0 +1,431 @@
+        pred_dG = np.trapz([_mean_du_dlambda(x) for x in results], lambda_schedule)
+        # TODO: refactor this to be a call to something in a generic free_energy.analysis module
+        #   to allow comparison / swapping between TI estimator, MBAR estimator, and others
+        stage_dGs[stage] = pred_dG
+
+        # collect derivative handlers
+        ghs = []
+        for lamb, result in zip(lambda_schedule, results):
+            bonded_du_dl, nonbonded_du_dl, grads_and_handles = result
+            ghs.append(grads_and_handles)
+
+        # update forcefield gradient handlers in-place
+        _update_combined_handle_and_grads(ghs, combined_handle_and_grads)
+
+    pred_rbfe = stage_dGs['complex'] - stage_dGs['solvent']
+    return pred_rbfe, combined_handle_and_grads, stage_results
+
+
+# TODO: define more flexible update rules here, rather than update parameters
+step_sizes = {
+    nonbonded.AM1CCCHandler: 1e-3,
+    nonbonded.LennardJonesHandler: 1e-3,
+    # ...
+}
+
+gradient_clip_thresholds = {
+    nonbonded.AM1CCCHandler: 0.001,
+    nonbonded.LennardJonesHandler: np.array([0.001, 0]),  # TODO: allow to update epsilon also?
+    # ...
+}
+
+
+def _clipped_update(gradient, step_size, clip_threshold):
+    """Compute an update based on current gradient
+        x[k+1] = x[k] + update
+
+    The gradient descent update would be
+        update = - step_size * grad(x[k]),
+
+    and to avoid instability, we clip the absolute values of all components of the update
+        update = - clip(step_size * grad(x[k]))
+
+    TODO: menu of other, fancier update functions
+    """
+    return - np.clip(step_size * gradient, -clip_threshold, clip_threshold)
+
+
+class ParameterUpdate:
+    def __init__(self, before, after, gradient, update):
+        self.before = before
+        self.after = after
+        self.gradient = gradient
+        self.update = update
+
+    # TODO: def __str__
+
+    def save(self, name='parameter_update.npz'):
+        """save numpy arrays to path/to/results/{name}.npz"""
+        parameter_update_path = path_to_results.joinpath(name)
+        print(f'saving parameter updates to {parameter_update_path}')
+        np.savez(
+            file=parameter_update_path,
+            before=self.before,
+            after=self.after,
+            gradient=self.gradient,
+            update=self.update,
+        )
+
+
+def _update_in_place(pred, grads, label,
+                     handle_types_to_update=[nonbonded.AM1CCCHandler, nonbonded.LennardJonesHandler]):
+    """
+    Notes
+    -----
+    * Currently hard-codes l1 loss -- may want to later break this up
+
+    TODO: check if it's too expensive to do out-of-place updates with a copy
+    """
+
+    # TODO: refactor loss definition out of this function
+    loss = np.abs(pred - label)
+    unit = "kJ/mol"
+
+    message = f"""
+    pred:   {pred:.3f} {unit}
+    target: {label:.3f} {unit} 
+    loss:   {loss:.3f} {unit}
+    """
```

Comment:
Would it be possible to include timings? Would be useful for us to assess speed/performance later on.

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/331/commits/89cbff0da504dad1efd20a0cd1c4ad637448b723